contracts/PillarStaking.sol

```diff
@@ -0,0 +1,271 @@
+        bool staked;
+        uint256 stakedAmount;
+        uint256 rewardAmount;
+    }
+
+    enum StakingState {
+        INITIALIZED,
+        STAKEABLE,
+        STAKED,
+        READY_FOR_UNSTAKE
+    }
+    StakingState state;
+
+    /* ========== EVENTS ========== */
+
+    event Staked(address indexed user, uint256 amount);
+    event Unstaked(address indexed user, uint256 amount);
+    event RewardsAllocated();
+    event RewardPaid(address indexed user, uint256 reward);
+    event ContractStateUpdated(StakingState newState);
+    event MinStakeAmountUpdated(uint256 newMinStake);
+    event MaxStakeAmountUpdated(uint256 newMaxStake);
+
+    /* ========== MODIFIERS ========== */
+
+    modifier whenInitialized() {
+        if (state != StakingState.INITIALIZED) revert OnlyWhenInitialized();
+        _;
+    }
+
+    modifier whenStakeable() {
+        if (state != StakingState.STAKEABLE) revert OnlyWhenStakeable();
+        _;
+    }
+
+    modifier whenStaked() {
+        if (state != StakingState.STAKED) revert OnlyWhenStaked();
+        _;
+    }
+
+    modifier whenReadyForUnstake() {
+        if (state != StakingState.READY_FOR_UNSTAKE)
+            revert OnlyWhenReadyForUnstake();
+        _;
+    }
+
+    modifier whenRewardsAllocated() {
+        if (!rewardsAllocated) revert RewardsNotAllocated();
+        _;
+    }
+
+    /* ========== CONSTRUCTOR ========== */
+
+    constructor(
+        address _stakingToken,
+        address _rewardToken,
+        uint256 _maxTotalStake
+    ) {
+        if (_stakingToken == address(0)) revert InvalidStakingToken();
+        if (_rewardToken == address(0)) revert InvalidRewardsToken();
+        if (_maxTotalStake > 0) maxTotalStake = _maxTotalStake;
+        stakingToken = IERC20(_stakingToken);
+        rewardToken = IERC20(_rewardToken);
+        setStateInitialized();
+    }
+
+    /* ========== MUTATIVE FUNCTIONS ========== */
+
+    function stake(uint256 _amount) external nonReentrant whenStakeable {
+        if (block.timestamp > stakingPeriod + STAKING_PERIOD)
+            revert("StakingPeriodPassed()");
+        if ((totalStaked + _amount) > maxTotalStake)
+            revert MaximumTotalStakeReached({
+                totalMaxStake: maxTotalStake,
+                currentStakedAmount: totalStaked,
+                remainingStakeableAmount: (maxTotalStake - totalStaked),
+                stakerAmount: _amount
+            });
+        if (_amount < minStake)
+            revert InvalidMinimumStake({minimumStakeAmount: minStake});
+        if (_amount > maxStake)
+            revert InvalidMaximumStake({maximumStakeAmount: maxStake});
+        if (_amount > maxStake - stakeholderData[msg.sender].stakedAmount)
+            revert StakeWouldBeGreaterThanMax();
+        if (stakingToken.balanceOf(msg.sender) < _amount)
+            revert InsufficientBalance();
+        totalStaked = totalStaked + _amount;
+        stakeholderData[msg.sender].stakedAmount =
```

Comment:
Please handle a scenario where a user can stake any amount between minStake and maxStaje multiple times, his cumulative totalStake shouldn't exceed the maxStake.

Reply:
That is covered in the revert above this:
```
if (_amount > maxStake - stakeholderData[msg.sender].stakedAmount)
            revert StakeWouldBeGreaterThanMax();
```

Reply:
Thanks. Pls add a specific tests for this.